File: samples/compose/app/src/main/kotlin/dev/getmilano/sample/ui/screens/PokemonScreen.kt
package dev.getmilano.sample.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import dev.getmilano.MilanoValue
import dev.getmilano.sample.environment.SampleEnvironment
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.URL

/**
 * One MilanoHost, two context sources: the app-wide shared context (the
 * trainer name) plus screen-specific values fetched from PokeAPI and
 * injected into this screen's builder. The document declares all four
 * keys; the gate validates them together at build.
 */
@Composable
fun PokemonScreen(environment: SampleEnvironment) {
    var screenContext by remember { mutableStateOf<Map<String, MilanoValue>?>(null) }
    var fetchFailure by remember { mutableStateOf<String?>(null) }

    // The screen owns its data: fetched before the document is built,
    // then handed to Milano as plain context values.
    LaunchedEffect(Unit) {
        try {
            val body =
                withContext(Dispatchers.IO) {
                    URL("https://pokeapi.co/api/v2/pokemon/pikachu").readText()
                }
            val json = JSONObject(body)
            val artwork =
                json
                    .getJSONObject("sprites")
                    .getJSONObject("other")
                    .getJSONObject("official-artwork")
                    .getString("front_default")
            screenContext =
                mapOf(
                    "pokemonName" to MilanoValue.StringValue(json.getString("name")),
                    "pokemonHeight" to MilanoValue.DoubleValue(json.getInt("height").toDouble()),
                    "pokemonWeight" to MilanoValue.DoubleValue(json.getInt("weight").toDouble()),
                    "pokemonImageUrl" to MilanoValue.StringValue(artwork),
                )
        } catch (error: Exception) {
            fetchFailure = error.toString()
        }
    }

    val context = screenContext
    val failure = fetchFailure
    when {
        context != null -> {
            DemoScreen(builder = environment.pokemonBuilder(context))
        }

        failure != null -> {
            Column(
                modifier = Modifier.fillMaxWidth().padding(24.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Text("Fetch failed", style = MaterialTheme.typography.titleMedium)
                Text(failure, style = MaterialTheme.typography.bodySmall)
            }
        }

        else -> {
            Column(
                modifier = Modifier.fillMaxWidth().padding(top = 120.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                CircularProgressIndicator()
                Text("Fetching from PokeAPI…")
            }
        }
    }
}

File: samples/swiftui/Sources/Screens/PokemonScreen.swift
import Foundation
import MilanoSDK
import SwiftUI

/// One MilanoHost, two context sources: the app-wide shared context (the
/// trainer name) plus screen-specific values fetched from PokeAPI and
/// injected into this screen's builder. The document declares all four
/// keys; the gate validates them together at build.
struct PokemonScreen: View {
    @State private var screenContext: [String: MilanoValue]?
    @State private var fetchFailure: String?

    var body: some View {
        content
            .navigationTitle("Pokemon")
    }

    @ViewBuilder
    private var content: some View {
        if let screenContext {
            ScrollView {
                MilanoHost(
                    builder: SampleEnvironment.shared.documentBuilder(
                        resource: "pokemon", screenContext: screenContext)
                ) {
                    ProgressView()
                } failure: { error in
                    Text(String(describing: error))
                        .font(.caption)
                        .padding()
                }
            }
        } else if let fetchFailure {
            VStack(spacing: 8) {
                Text("Fetch failed").font(.headline)
                Text(fetchFailure)
                    .font(.caption)
                    .foregroundColor(.secondary)
            }
            .padding()
        } else {
            ProgressView("Fetching from PokeAPI…")
                .task { await fetch() }
        }
    }

    /// The screen owns its data: fetched before the document is built,
    /// then handed to Milano as plain context values.
    private func fetch() async {
        do {
            let url = URL(string: "https://pokeapi.co/api/v2/pokemon/pikachu")!
            let (data, _) = try await URLSession.shared.data(from: url)
            guard
                let object = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                let name = object["name"] as? String,
                let height = object["height"] as? Int,
                let weight = object["weight"] as? Int,
                let sprites = object["sprites"] as? [String: Any],
                let other = sprites["other"] as? [String: Any],
                let artwork = other["official-artwork"] as? [String: Any],
                let imageUrl = artwork["front_default"] as? String
            else {
                fetchFailure = "unexpected PokeAPI payload"
                return
            }
            screenContext = [
                "pokemonName": .string(name),
                "pokemonHeight": .double(Double(height)),
                "pokemonWeight": .double(Double(weight)),
                "pokemonImageUrl": .string(imageUrl)
            ]
        } catch {
            fetchFailure = String(describing: error)
        }
    }
}
